File: flutter-plugin/ios/Classes/ConsentPlugin.swift
import Flutter
import UIKit

/// Flutter platform channel plugin for the Consent SDK on iOS.
public class ConsentPlugin: NSObject, FlutterPlugin {

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: "consent_sdk",
            binaryMessenger: registrar.messenger()
        )
        let instance = ConsentPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "initialize":
            handleInitialize(call, result: result)
        case "showConsent":
            handleShowConsent(result: result)
        case "hideConsent":
            handleHideConsent(result: result)
        case "setLanguage":
            handleSetLanguage(call, result: result)
        case "getConsentStatus":
            handleGetConsentStatus(result: result)
        case "resetConsent":
            handleResetConsent(result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func handleInitialize(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let apiKey = args["apiKey"] as? String,
              let baseUrl = args["baseUrl"] as? String else {
            result(FlutterError(code: "INIT_ERROR", message: "apiKey and baseUrl are required", details: nil))
            return
        }

        let language = args["language"] as? String ?? "en"
        let minimumAge = args["minimumAge"] as? Int ?? 16
        let timeoutMs = args["timeoutMs"] as? Double ?? 15000
        let retryCount = args["retryCount"] as? Int ?? 2

        var theme = ConsentTheme.light
        if let themeDict = args["theme"] as? [String: Any] {
            theme = parseTheme(themeDict)
        }

        let config = ConsentConfig(
            apiKey: apiKey,
            baseUrl: baseUrl,
            language: language,
            theme: theme,
            minimumAge: minimumAge,
            timeoutInterval: timeoutMs / 1000.0,
            retryCount: retryCount
        )

        DispatchQueue.main.async {
            ConsentSDK.shared.initialize(config: config)
            result(nil)
        }
    }

    private func handleShowConsent(result: @escaping FlutterResult) {
        DispatchQueue.main.async {
            let sdkResult = ConsentSDK.shared.showConsent()
            switch sdkResult {
            case .success:
                result(nil)
            case .error(let message, let code):
                result(FlutterError(code: code.rawValue, message: message, details: nil))
            }
        }
    }

    private func handleHideConsent(result: @escaping FlutterResult) {
        DispatchQueue.main.async {
            ConsentSDK.shared.hideConsent()
            result(nil)
        }
    }

    private func handleSetLanguage(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let langCode = args["langCode"] as? String else {
            result(FlutterError(code: "LANG_ERROR", message: "langCode is required", details: nil))
            return
        }

        DispatchQueue.main.async {
            ConsentSDK.shared.setLanguage(langCode)
            result(nil)
        }
    }

    private func handleGetConsentStatus(result: @escaping FlutterResult) {
        DispatchQueue.main.async {
            result(ConsentSDK.shared.getConsentStatus())
        }
    }

    private func handleResetConsent(result: @escaping FlutterResult) {
        DispatchQueue.main.async {
            ConsentSDK.shared.resetConsent()
            result(nil)
        }
    }

    private func parseTheme(_ dict: [String: Any]) -> ConsentTheme {
        return ConsentTheme(
            primaryColor: Color(hex: dict["primaryColor"] as? String ?? "#6200EE"),
            backgroundColor: Color(hex: dict["backgroundColor"] as? String ?? "#FFFFFF"),
            textColor: Color(hex: dict["textColor"] as? String ?? "#212121"),
            buttonRadius: CGFloat(dict["buttonRadius"] as? Double ?? 8),
            fontFamily: dict["fontFamily"] as? String ?? "System",
            checkboxColor: Color(hex: dict["checkboxColor"] as? String ?? "#6200EE"),
            buttonTextColor: Color(hex: dict["buttonTextColor"] as? String ?? "#FFFFFF"),
            overlayColor: Color(hex: dict["overlayColor"] as? String ?? "#80000000"),
            errorColor: Color(hex: dict["errorColor"] as? String ?? "#B00020")
        )
    }
}

File: flutter-plugin/android/src/main/kotlin/com/niceoneconsent/sdk/flutter/ConsentPlugin.kt
package com.niceoneconsent.sdk.flutter

import android.app.Activity
import androidx.annotation.NonNull
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import com.niceoneconsent.sdk.android.ConsentSDK
import com.niceoneconsent.sdk.models.ConsentConfig
import com.niceoneconsent.sdk.models.ConsentTheme
import com.niceoneconsent.sdk.models.ConsentResult

/**
 * Flutter platform channel plugin for the Consent SDK on Android.
 */
class ConsentPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {

    private lateinit var channel: MethodChannel
    private var activity: Activity? = null

    override fun onAttachedToEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(binding.binaryMessenger, "consent_sdk")
        channel.setMethodCallHandler(this)
    }

    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    // ActivityAware
    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivity() {
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    // MethodCallHandler
    override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
        when (call.method) {
            "initialize" -> handleInitialize(call, result)
            "showConsent" -> handleShowConsent(result)
            "hideConsent" -> handleHideConsent(result)
            "setLanguage" -> handleSetLanguage(call, result)
            "getConsentStatus" -> handleGetConsentStatus(result)
            "resetConsent" -> handleResetConsent(result)
            else -> result.notImplemented()
        }
    }

    private fun handleInitialize(call: MethodCall, result: Result) {
        try {
            val apiKey = call.argument<String>("apiKey") ?: ""
            val baseUrl = call.argument<String>("baseUrl") ?: ""
            val language = call.argument<String>("language") ?: "en"
            val minimumAge = call.argument<Int>("minimumAge") ?: 16
            val timeoutMs = call.argument<Int>("timeoutMs")?.toLong() ?: 15000L
            val retryCount = call.argument<Int>("retryCount") ?: 2

            val themeMap = call.argument<Map<String, Any>>("theme")
            val theme = parseTheme(themeMap)

            val config = ConsentConfig(
                apiKey = apiKey,
                baseUrl = baseUrl,
                language = language,
                theme = theme,
                minimumAge = minimumAge,
                timeoutMs = timeoutMs,
                retryCount = retryCount
            )

            ConsentSDK.initialize(config)
            result.success(null)
        } catch (e: Exception) {
            result.error("INIT_ERROR", "Failed to initialize: ${e.message}", null)
        }
    }

    private fun handleShowConsent(result: Result) {
        val currentActivity = activity
        if (currentActivity == null) {
            result.error("NO_ACTIVITY", "No activity available", null)
            return
        }

        when (val sdkResult = ConsentSDK.showConsent(currentActivity)) {
            is ConsentResult.Success -> result.success(null)
            is ConsentResult.Error -> result.error(sdkResult.code.name, sdkResult.message, null)
        }
    }

    private fun handleHideConsent(result: Result) {
        ConsentSDK.hideConsent()
        result.success(null)
    }

    private fun handleSetLanguage(call: MethodCall, result: Result) {
        val langCode = call.argument<String>("langCode") ?: "en"
        ConsentSDK.setLanguage(langCode)
        result.success(null)
    }

    private fun handleGetConsentStatus(result: Result) {
        result.success(ConsentSDK.getConsentStatus())
    }

    private fun handleResetConsent(result: Result) {
        ConsentSDK.resetConsent()
        result.success(null)
    }

    private fun parseTheme(themeMap: Map<String, Any>?): ConsentTheme {
        if (themeMap == null) return ConsentTheme.Light

        return ConsentTheme(
            primaryColor = themeMap["primaryColor"] as? String ?: "#6200EE",
            backgroundColor = themeMap["backgroundColor"] as? String ?: "#FFFFFF",
            textColor = themeMap["textColor"] as? String ?: "#212121",
            buttonRadius = (themeMap["buttonRadius"] as? Double)?.toFloat() ?: 8f,
            fontFamily = themeMap["fontFamily"] as? String ?: "System",
            checkboxColor = themeMap["checkboxColor"] as? String ?: "#6200EE",
            buttonTextColor = themeMap["buttonTextColor"] as? String ?: "#FFFFFF",
            overlayColor = themeMap["overlayColor"] as? String ?: "#80000000",
            errorColor = themeMap["errorColor"] as? String ?: "#B00020",
            cardElevation = (themeMap["cardElevation"] as? Double)?.toFloat() ?: 8f,
            secondaryButtonColor = themeMap["secondaryButtonColor"] as? String ?: "#757575",
            secondaryButtonTextColor = themeMap["secondaryButtonTextColor"] as? String ?: "#FFFFFF",
            inputBorderColor = themeMap["inputBorderColor"] as? String ?: "#BDBDBD",
            inputBackgroundColor = themeMap["inputBackgroundColor"] as? String ?: "#F5F5F5",
            mandatoryBadgeColor = themeMap["mandatoryBadgeColor"] as? String ?: "#FF9800"
        )
    }
}
